src/lib/comms/sol-mqtt.c

```diff
@@ -0,0 +1,579 @@
+sol_mqtt_publish(struct sol_mqtt *mqtt, struct sol_mqtt_message *message)
+{
+    int r;
+
+    CHECK_INIT(-EINVAL);
+    SOL_NULL_CHECK(mqtt, -EINVAL);
+    SOL_NULL_CHECK(message, -EINVAL);
+
+    r = mosquitto_publish(mqtt->mosq, &message->id, message->topic, message->payload_len, message->payload,
+            (int) message->qos, message->retain);
+
+    if (r != MOSQ_ERR_SUCCESS) {
+        SOL_WRN("Unable to publish to '%s'", message->topic);
+        return -EINVAL;
+    }
+
+    if (!mqtt->socket_write) {
+        mqtt->socket_write = sol_fd_add(mqtt->socket_fd, SOL_FD_FLAGS_OUT, sol_mqtt_event_loop, mqtt);
+        SOL_NULL_CHECK(mqtt->socket_write, -EINVAL);
+    }
+
+    return 0;
+}
+
+SOL_API int
+sol_mqtt_subscribe(struct sol_mqtt *mqtt, char *topic, size_t topic_len, sol_mqtt_qos qos)
+{
+    int r;
+    char *s_topic;
+
+    CHECK_INIT(-EINVAL);
+    SOL_NULL_CHECK(mqtt, -EINVAL);
+    SOL_NULL_CHECK(topic, -EINVAL);
+
+    s_topic = strndupa(topic, topic_len);
+
+    r = mosquitto_subscribe(mqtt->mosq, NULL, s_topic, (int) qos);
+    if (r != MOSQ_ERR_SUCCESS) {
+        SOL_WRN("Unable to subscribe to '%s'", topic);
+        return -EINVAL;
+    }
+
+    return 0;
+}
+
+SOL_API struct sol_mqtt_message *
+sol_mqtt_message_new(char *topic, size_t topic_len, void *payload, size_t payload_len,
+        sol_mqtt_qos qos, bool retain)
+{
+    struct sol_mqtt_message *message;
+
+    SOL_NULL_CHECK(topic, NULL);
+    SOL_NULL_CHECK(payload, NULL);
+
+    message = calloc(1, sizeof(struct sol_mqtt_message));
+    SOL_NULL_CHECK(message, NULL);
+
+    message->topic = calloc(1, topic_len + 1);
+    SOL_NULL_CHECK_GOTO(message->topic, topic_error);
+    memcpy(message->topic, topic, topic_len);
+
+    message->payload = sol_util_memdup(payload, payload_len);
+    SOL_NULL_CHECK_GOTO(message->topic, payload_error);
+
+    message->payload_len = payload_len;
+
+    message->qos = qos;
+    message->retain = retain;
+
+    return message;
+
+payload_error:
+    free(message->topic);
+topic_error:
+    free(message);
+
+    return NULL;
+}
+
+SOL_API struct sol_mqtt_message *
+sol_mqtt_message_copy(struct sol_mqtt_message *message)
+{
+    struct sol_mqtt_message *m_copy;
+    SOL_NULL_CHECK(message, NULL);
+
+    m_copy = sol_mqtt_message_new(message->topic, strlen(message->topic), message->payload,
+            message->payload_len, message->qos, message->retain);
+
+    SOL_NULL_CHECK(m_copy, NULL);
+
+    m_copy->id = message->id;
+
+    return m_copy;
+}
+
+SOL_API void
+sol_mqtt_message_free(struct sol_mqtt_message *message) {
+    if (!message)
+        return;
+
+    free(message->topic);
+    free(message->payload);
+    free(message);
+}
+
+SOL_API size_t
+sol_mqtt_get_topic(const struct sol_mqtt_message *message, char *buf, size_t size)
+{
+    size_t s;
+
+    SOL_NULL_CHECK(message, 0);
+    SOL_NULL_CHECK(buf, 0);
+
+    s = strlen(message->topic);
+    if (s > size)
+        s = size;
+
+    strncpy(buf, message->topic, s);
```

Comment:
well, whoever gets it, will get its length, also, ensuring \0 could possibly trim out the last byte since we don't control the dst buffer. Anyway, will change everywhere just to use/expect a NULL terminated string.